From 382d6acbebbe0dd58d53b66ac83847f37baa5488 Mon Sep 17 00:00:00 2001
From: Yongqiang Sun <yongqiang.sun@amd.com>
Date: Wed, 4 May 2016 16:04:13 -0400
Subject: [PATCH 0478/1722] drm/amd/dal: Add break when wait flip interrupt
 time out.

Issue:
In case of headless switching DP passive dongle to DP display,
DP display not light up.
cause:
Flip interrupt is damaged by verify link cap due to incorrect clock source.
System stucks at waiting flip done before disable flip interrupt.
Work around:
Change time out processing, use WARN_ON instead of BUG_ON,
in case of time out, add break to exit while loop.

Signed-off-by: Yongqiang Sun <yongqiang.sun@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
index 2f1c069..bc9ec07 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
@@ -2009,7 +2009,20 @@ static void wait_while_pflip_status(struct amdgpu_device *adev,
 										__func__, acrtc->crtc_id,
 										acrtc,
 										acrtc->pflip_status);
-			BUG_ON(1);
+
+			/* we do not expect to hit this case except on Polaris with PHY PLL
+			 * 1. DP to HDMI passive dongle connected
+			 * 2. unplug (headless)
+			 * 3. plug in DP
+			 * 3a. on plug in, DP will try verify link by training, and training
+			 * would disable PHY PLL which HDMI rely on to drive TG
+			 * 3b. this will cause flip interrupt cannot be generated, and we
+			 * exit when timeout expired.  however we do not have code to clean
+			 * up flip, flip clean up will happen when the address is written
+			 * with the restore mode change
+			 */
+			WARN_ON(1);
+			break;
 		}
 	}
 
-- 
2.7.4

